### ISSUE TITLE
bug: S3 Multipart Upload 500

### ISSUE THREAD

**Issue by leonardosul** [localstack/localstack#7104]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

When using [AWS Data Wrangler](https://github.com/aws/aws-sdk-pandas) and the [awswrangler.s3.to_parquet](https://aws-sdk-pandas.readthedocs.io/en/2.17.0/stubs/awswrangler.s3.to_parquet.html#awswrangler.s3.to_parquet) method I get 500 errors from Localstack:

`KeyError: 'nzSOyWi9bGaavSXqEpB6Tcad8TsUzRRVEa4092avHhN90aI82cvyUN4w'
2022-10-27T10:30:04.444062181Z 2022-10-27T10:30:03.833  INFO --- [   asgi_gw_2] localstack.request.aws     : AWS s3.UploadPart => 500 (500)
2022-10-27T10:30:12.452558051Z 2022-10-27T10:30:12.226 ERROR --- [uest_thread)] werkzeug                   : Error on request:
2022-10-27T10:30:12.452600321Z Traceback (most recent call last):
2022-10-27T10:30:12.452608531Z   File "/opt/code/localstack/.venv/lib/python3.10/site-packages/werkzeug/serving.py", line 335, in run_wsgi
2022-10-27T10:30:12.452613909Z     execute(self.server.app)
2022-10-27T10:30:12.452618714Z   File "/opt/code/localstack/.venv/lib/python3.10/site-packages/werkzeug/serving.py", line 322, in execute
2022-10-27T10:30:12.452635778Z     application_iter = app(environ, start_response)
2022-10-27T10:30:12.452640541Z   File "/opt/code/localstack/.venv/lib/python3.10/site-packages/moto/moto_server/werkzeug_app.py", line 249, in __call__
2022-10-27T10:30:12.452644707Z     return backend_app(environ, start_response)
2022-10-27T10:30:12.452648715Z   File "/opt/code/localstack/.venv/lib/python3.10/site-packages/flask/app.py", line 2091, in __call__
2022-10-27T10:30:12.452652738Z     return self.wsgi_app(environ, start_response)
2022-10-27T10:30:12.452656622Z   File "/opt/code/localstack/.venv/lib/python3.10/site-packages/flask/app.py", line 2076, in wsgi_app
2022-10-27T10:30:12.452660683Z     response = self.handle_exception(e)
2022-10-27T10:30:12.452664798Z   File "/opt/code/localstack/.venv/lib/python3.10/site-packages/flask_cors/extension.py", line 165, in wrapped_function
2022-10-27T10:30:12.452669211Z     return cors_after_request(app.make_response(f(*args, **kwargs)))
2022-10-27T10:30:12.452673159Z   File "/opt/code/localstack/.venv/lib/python3.10/site-packages/flask/app.py", line 2073, in wsgi_app
2022-10-27T10:30:12.452677174Z     response = self.full_dispatch_request()
2022-10-27T10:30:12.452681007Z   File "/opt/code/localstack/.venv/lib/python3.10/site-packages/flask/app.py", line 1519, in full_dispatch_request
2022-10-27T10:30:12.452685042Z     rv = self.handle_user_exception(e)
2022-10-27T10:30:12.452688967Z   File "/opt/code/localstack/.venv/lib/python3.10/site-packages/flask_cors/extension.py", line 165, in wrapped_function
2022-10-27T10:30:12.452693026Z     return cors_after_request(app.make_response(f(*args, **kwargs)))
2022-10-27T10:30:12.452696986Z   File "/opt/code/localstack/.venv/lib/python3.10/site-packages/flask/app.py", line 1517, in full_dispatch_request
2022-10-27T10:30:12.452700984Z     rv = self.dispatch_request()
2022-10-27T10:30:12.452704881Z   File "/opt/code/localstack/.venv/lib/python3.10/site-packages/flask/app.py", line 1503, in dispatch_request
2022-10-27T10:30:12.452709096Z     return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)
2022-10-27T10:30:12.452713187Z   File "/opt/code/localstack/localstack/utils/patch.py", line 38, in proxy
2022-10-27T10:30:12.452717192Z     return new(target, *args, **kwargs)
2022-10-27T10:30:12.452721879Z   File "/opt/code/localstack/localstack/utils/aws/request_context.py", line 174, in convert_to_flask_response_call
2022-10-27T10:30:12.452726272Z     return fn(*args, **kwargs)
2022-10-27T10:30:12.452730309Z   File "/opt/code/localstack/.venv/lib/python3.10/site-packages/moto/core/utils.py", line 110, in __call__
2022-10-27T10:30:12.452734362Z     result = self.callback(request, request.url, dict(request.headers))
2022-10-27T10:30:12.452738319Z   File "/opt/code/localstack/.venv/lib/python3.10/site-packages/moto/utilities/aws_headers.py", line 60, in _wrapper
2022-10-27T10:30:12.452742452Z     response = f(*args, **kwargs)
2022-10-27T10:30:12.452746316Z   File "/opt/code/localstack/.venv/lib/python3.10/site-packages/moto/s3/responses.py", line 1123, in key_response
2022-10-27T10:30:12.452754993Z     response = self._key_response(request, full_url, self.headers)
2022-10-27T10:30:12.452759269Z   File "/opt/code/localstack/.venv/lib/python3.10/site-packages/moto/s3/responses.py", line 1214, in _key_response
2022-10-27T10:30:12.452763345Z     return self._key_response_put(request, body, bucket_name, query, key_name)
2022-10-27T10:30:12.452767296Z   File "/opt/code/localstack/localstack/utils/patch.py", line 38, in proxy
2022-10-27T10:30:12.452771327Z     return new(target, *args, **kwargs)
2022-10-27T10:30:12.452775277Z   File "/opt/code/localstack/localstack/services/s3/s3_starter.py", line 191, in s3_key_response_put
2022-10-27T10:30:12.452779415Z     result = fn(request, body, bucket_name, query, key_name, *args, **kwargs)
2022-10-27T10:30:12.452783243Z   File "/opt/code/localstack/.venv/lib/python3.10/site-packages/moto/s3/responses.py", line 1379, in _key_response_put
2022-10-27T10:30:12.452787614Z     key = self.backend.upload_part(
2022-10-27T10:30:12.452791684Z   File "/opt/code/localstack/.venv/lib/python3.10/site-packages/moto/s3/models.py", line 1997, in upload_part
2022-10-27T10:30:12.452795714Z     multipart = bucket.multiparts[multipart_id]
2022-10-27T10:30:12.452799596Z KeyError: 'nzSOyWi9bGaavSXqEpB6Tcad8TsUzRRVEa4092avHhN90aI82cvyUN4w'`

### Expected Behavior

The multipart upload completes as expected. The awswrangler package works fine when running against a real S3 bucket.

### How are you starting LocalStack?

With a docker-compose file

#### How are you starting localstack (e.g., `bin/localstack` command, arguments, or `docker-compose.yml`)

    docker run localstack/localstack:1.2.0 ( we use docker-compose)

#### Client commands (e.g., AWS SDK code snippet, or sequence of "awslocal" commands)

   Create an S3 bucket via any means (I use Terraform)

Run load a Pandas DataFrame from a source, such as a database. Try to write the DataFrame to S3 using awswrangler:

`            wr.s3.to_parquet(
                df=df_chunk,
                path=self._build_path(),
                dataset=True,
                mode="overwrite_partitions",
                s3_additional_kwargs=self.s3_extra_args,
            )
`

### Environment

```markdown
- OS: Host is MacOS, but we use docker compose for all services. The upload service is Debian
- LocalStack:1.2.0
```

### Anything else?

Single file uploads seem to work fine when using awswrangler.

**Comment by leonardosul** [comment]
Looks to be related to this: https://github.com/spulec/moto/pull/5589 which has been patched in upstream moto

**Comment by bblommers** [comment]
Can you reproduce this with Localstack 1.3.0 @leonardosul?

(Note: I'm not a LS engineer, but a Moto maintainer - to me it looks like the fix should be part of LS already. The bug was (should be?) fixed as of Moto 4.0.9, which became part of LS with #6991 - and the 1.3.0 release happened two days after that PR was merged.)

**Comment by ashin-czi** [comment]
Wanted to chime in to say that I am having similar issues with trying to upload a 7.4GB binary file

Tried LS 1.3.1 and 1.3.0 on Docker and I keep having the same issues with running out of memory (tried 8GB to 16GB memory limits) when trying to upload a large file with the `aws s3 sync` command.

It seems like right when all of the parts are uploaded, the memory increases to the point of crashing the container. Smaller files seem to upload fine.

Happy to provide additional details if needed!

**Comment by bentsku** [comment]
Hello @leonardosul and thanks for your report. It seems your issue is related to the use of the `S3_DIR` config flag, as we do not support Multipart Uploads with that variable yet. If you remove this flag, your issue will be solved. I am not sure what is your use case, and could it be covered by persistence (https://docs.localstack.cloud/references/persistence-mechanism/)?
I will close the issue as we have another report highlighting the problem with `S3_DIR` here: #8263. You will be able to follow the development of the fix there.

@ashin-czi, you are right, there is an issue with memory usage when completing a Multipart upload, as the underlying library will concatenate every part in memory before saving it to file. You will be able to follow the fix for this issue here: #6845. We're currently working on it, and it should fix the issue with a config flag.
